src/main/java/net/nextencia/rrdiagram/BNFProcessor.java

```diff
@@ -0,0 +1,262 @@
+     * reference file @ content/preview/api/ysql/syntax_resources/grammar_diagrams.md
+     * ysql/ycql is at depth 4 (from content)
+     */
+
+    String path = "";
+    
+    if (depth < 4) {
+      path = "/" + version + "/api/" + api + "/";
+    }
+
+    // for every level deeper from ysql/ycql, add a ../
+    switch(depth) {
+      case 5: path = "../"; break;
+      case 6: path = "../../"; break;
+      case 7: path = "../../../"; break;
+      case 8: path = "../../../../"; break;
+      case 9: path = "../../../../../"; break;
+    }
+
+    if (api.equals("ysql")) {
+      return path + "syntax_resources/grammar_diagrams";
+    } else {
+      return path + "grammar_diagrams";
+    }
+  }
+
+  public String getDiagram(List<Rule> targetRules , List<Rule> localRules, int depth) {
+    StringBuilder sb = new StringBuilder();
+    try {
+      final Set<String> localRefs = new HashSet<String>();
+
+      // current rules
+      for (Rule rule : targetRules) {
+        localRefs.add(rule.getName());
+      }
+
+      // other rules in the same page
+      for (Rule rule : localRules) {
+        localRefs.add(rule.getName());
+      }
+
+      String globalPrefix = getGlobalPrefix(depth);
+      GrammarToRRDiagram.RuleLinkProvider ruleLinkProvider =
+        new GrammarToRRDiagram.RuleLinkProvider() {
+          @Override
+          public String getLink(String ruleName) {
+            String linkName = "#" + ruleName.replaceAll("_", "-");
+            if (!localRefs.contains(ruleName)) {
+              linkName = globalPrefix + linkName;
+            }
+            return linkName;
+          }
+        };
+
+      for (Rule rule : targetRules) {
+        sb.append("#### " + rule.getName());
+        sb.append("\n\n");
+        GrammarToRRDiagram diagram_builder = new GrammarToRRDiagram();
+        diagram_builder.setRuleLinkProvider(ruleLinkProvider);
+        diagram_builder.setRuleConsideredAsLineBreak(Utils.lineBreakRule);
+        RRDiagram diagram = diagram_builder.convert(rule);
+        String svg_string = new RRDiagramToSVG().convert(diagram);
+        sb.append(svg_string);
+        sb.append("\n\n");
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+    return sb.toString();
+  }
+
+  public Set<String> getRuleNames() {
+    Set<String> ruleNames = new HashSet<String>();
+    ruleNames.add(Utils.lineBreakRule);
+    for (Rule rule : grammar.getRules()) {
+      if (!ruleNames.add(rule.getName())) {
+        logger.warning("Rule defined multiple times: " + rule.getName());
+      }
+    }
+
+    return ruleNames;
+  }
+
+  public void checkGrammar() {
+    Set<String> ruleNames = getRuleNames();
+
+    for (Rule rule : grammar.getRules()) {
+      Set<String> refs = rule.getUndefinedRuleRefs(ruleNames);
+      if (!refs.isEmpty()) {
+        logger.warning("Undefined rules referenced in rule '" + rule.getName() + "': " + refs.toString());
+      }
+    }
+  }
```

Comment:
This is copy-paste from Main.java.

Reply:
Most of the code is copy-paste from the Main.java file. This way we can ensure that the diagram generation is consistent with the previous flow.